File: v1/src/main/java/com/api/v1/medical_slots/utils/MedicalSlotFinderImpl.java
package com.api.v1.medical_slots.utils;

import com.api.v1.common.ObjectId;
import com.api.v1.doctors.Doctor;
import com.api.v1.medical_slots.MedicalSlot;
import com.api.v1.medical_slots.MedicalSlotFinder;
import com.api.v1.medical_slots.domain.MedicalSlotCrudRepository;
import com.api.v1.medical_slots.exceptions.InaccessibleMedicalSlotException;
import com.api.v1.medical_slots.exceptions.MedicalSlotNotFoundException;
import jakarta.validation.constraints.NotNull;
import org.springframework.stereotype.Component;

import java.util.Optional;

@Component
public final class MedicalSlotFinderImpl implements MedicalSlotFinder {

    private final MedicalSlotCrudRepository crudRepository;

    public MedicalSlotFinderImpl(MedicalSlotCrudRepository crudRepository) {
        this.crudRepository = crudRepository;
    }

    @Override
    public MedicalSlot findById(@ObjectId String id) {
        return crudRepository
                .findById(id)
                .orElseThrow(() -> new MedicalSlotNotFoundException(id));
    }

    @Override
    public MedicalSlot findByIdAndDoctor(@ObjectId String id, @NotNull Doctor doctor) {
        MedicalSlot foundSlot = findById(id);
        if (!foundSlot.doctor().id().equals(doctor.id())) {
            throw new InaccessibleMedicalSlotException(id, doctor.licenseNumber());
        }
        return foundSlot;
    }
}

File: v2/src/main/kotlin/com/api/v2/medical_slots/utils/MedicalSlotFinderImpl.kt
package com.api.v2.medical_slots.utils

import com.api.v2.doctors.Doctor
import com.api.v2.medical_slots.MedicalSlot
import com.api.v2.medical_slots.MedicalSlotFinder
import com.api.v2.medical_slots.domain.MedicalSlotCrudRepository
import com.api.v2.medical_slots.exceptions.InaccessibleMedicalSlotException
import com.api.v2.medical_slots.exceptions.MedicalSlotNotFoundException
import org.springframework.stereotype.Service

@Service
class MedicalSlotFinderImpl(
    private val crudRepository: MedicalSlotCrudRepository
): MedicalSlotFinder {

    override fun findById(id: String): MedicalSlot {
        val foundSlot = crudRepository.findById(id)
        if (foundSlot.isEmpty) {
            throw MedicalSlotNotFoundException(id)
        }
        return foundSlot.get()
    }

    override fun findByIdAndDoctor(id: String, doctor: Doctor): MedicalSlot {
        val foundSlot = findById(id)
        if (foundSlot.doctor.id != doctor.id) {
            throw InaccessibleMedicalSlotException(id, doctor.licenseNumber)
        }
        return foundSlot
    }
}